src/transaction/log_recovery_analysis.cpp

```diff
@@ -2241,20 +2231,105 @@ log_recovery_analysis_from_transaction_table_snapshot (THREAD_ENTRY *thread_p,
 
   lr.advance_when_does_not_fit (sizeof (log_rec_trantable_snapshot));
   const log_rec_trantable_snapshot log_rec = lr.reinterpret_copy_and_add_align<log_rec_trantable_snapshot> ();
-  std::unique_ptr<char []> snapshot_data_buf { new char[log_rec.length] };
+  std::unique_ptr<char []> snapshot_data_buf = std::make_unique<char []> (static_cast<size_t> (log_rec.length));
   lr.copy_from_log (snapshot_data_buf.get (), log_rec.length);
 
+  snapshot_lsa = log_rec.snapshot_lsa;
+
   cubpacking::unpacker unpacker;
   unpacker.set_buffer (snapshot_data_buf.get (), log_rec.length);
 
-  cublog::checkpoint_info chkpt_info;
   chkpt_info.unpack (unpacker);
 
+  return NO_ERROR;
+}
+
+static void
+log_recovery_build_mvcc_table_from_trantable_snapshot (THREAD_ENTRY *thread_p)
+{
+  assert (is_passive_transaction_server ());
+  assert (LOG_CS_OWN_WRITE_MODE (thread_p));
+
+  MVCCID smallest_mvccid = std::numeric_limits<MVCCID>::max ();
+  MVCCID largest_mvccid = std::numeric_limits<MVCCID>::min ();
+  std::set<MVCCID> present_mvccids;
+  for (int i = 0; i < log_Gl.trantable.num_total_indices; ++i)
+    {
+      if (i != LOG_SYSTEM_TRAN_INDEX)
+	{
+	  const log_tdes *const tdes = log_Gl.trantable.all_tdes[i];
+	  if (tdes != nullptr && tdes->trid != NULL_TRANID)
+	    {
+	      if (tdes->mvccinfo.id < smallest_mvccid)
+		{
+		  smallest_mvccid = tdes->mvccinfo.id;
+		}
+	      if (tdes->mvccinfo.id > largest_mvccid)
+		{
+		  largest_mvccid = tdes->mvccinfo.id;
+		}
+	      present_mvccids.insert (tdes->mvccinfo.id);
+	    }
+	}
+    }
+  log_Gl.hdr.mvcc_next_id = smallest_mvccid;
```

Comment:
Do we need to keep log_Gl.hdr.mvcc_next_id == smallest_mvccid during the loop which "commits" missing MVCCIDs or is needed only for reset_start_mvccid () ?

Reply:
It is needed for `mvcctable::reset_start_mvccid`. But I am not sure about the order in relation with "committing" missing MVCCID's.